diff --git a/meson.build b/meson.build
index 645c6ac1d..724ad0a1d 100644
--- a/meson.build
+++ b/meson.build
@@ -760,7 +760,7 @@ endif
 # force patched version of libcyaml from subproject
 # see https://github.com/tlsa/libcyaml/issues/150
 cyaml_dep = dependency(
-  'libcyaml', required: false, version: '>=99.1.0',
+  'libcyaml', required: false, version: '>=1.1.0',
   static: all_static)
 if not cyaml_dep.found()
   cyaml_proj = subproject('libcyaml')
